src/Disciplina/Crypto/MerkleTree.hs

```diff
@@ -0,0 +1,232 @@
+import Data.Tree as Tree (Tree(Node), drawTree)
+
+import qualified Data.Foldable as F (Foldable (..))
+import qualified Data.ByteString.Builder.Extra as Builder
+import qualified Data.ByteString.Lazy as LBS
+
+-- | Data type for root of sized merkle tree.
+data MerkleSignature a = MerkleSignature
+    { smrHash :: Hash LBS.ByteString  -- ^ returns root 'Hash' of Merkle Tree
+    , smrSize :: Word8 -- ^ size of root node,
+                       -- size is defined as number of leafs in this subtree
+    } deriving (Show, Eq, Ord, Generic, Serialise, Functor, Typeable)
+
+
+data MerkleTree a
+  = MerkleEmpty
+  | MerkleTree !(MerkleNode a)
+  deriving (Eq, Show, Generic)
+
+
+instance Foldable MerkleTree where
+    foldMap _ MerkleEmpty    = mempty
+    foldMap f (MerkleTree n) = F.foldMap f n
+
+    null MerkleEmpty      = True
+    null _                = False
+
+    length MerkleEmpty    = 0
+    length (MerkleTree n) = fromIntegral (smrSize (mRoot n))
+
+deriving instance Container (MerkleTree a)
+
+type LeafIndex = Int
+
+data MerkleNode a
+    = MerkleBranch { mRoot  :: !(MerkleSignature a)
+                   , mLeft  :: !(MerkleNode a)
+                   , mRight :: !(MerkleNode a)}
+    | MerkleLeaf { mRoot  :: !(MerkleSignature a)
+                 , mIndex :: !LeafIndex
+                 , mVal   :: !a}
+    deriving (Eq, Show, Functor, Generic)
+
+instance Foldable MerkleNode where
+  foldMap f x = case x of
+    MerkleLeaf {mVal} -> f mVal
+    MerkleBranch {mLeft, mRight} -> F.foldMap f mLeft `mappend` F.foldMap f mRight
+
+mkLeaf :: HasHash a => (LeafIndex, a) -> MerkleNode a
+mkLeaf (i, a) = MerkleLeaf
+    { mVal  = a
+    , mIndex = i
+    , mRoot = MerkleSignature (unsafeHash a) -- unsafeHash since we need to hash to ByteString
+                              1 -- size of leaf node is 1
+    }
+
+mkBranch :: MerkleNode a -> MerkleNode a -> MerkleNode a
+mkBranch l r =
+    MerkleBranch
+    { mLeft  = l
+    , mRight = r
+    , mRoot  = mkBranchRootHash (mRoot l) (mRoot r)
+    }
+
+mkBranchRootHash :: MerkleSignature a -- ^ left merkle root
+                 -> MerkleSignature a -- ^ right merkle root
+                 -> MerkleSignature a
+mkBranchRootHash (MerkleSignature (AbstractHash hl) sl)
+                 (MerkleSignature (AbstractHash hr) sr)
+  = MerkleSignature
+     (hash $ toLazyByteString $ mconcat
+        [ byteString (one sl)
+        , byteString (one sr)
+        , byteString (convert hl)
+        , byteString (convert hr) ])
+     (sl + sr)
+  where
+    toLazyByteString :: Builder -> LBS.ByteString
+    toLazyByteString = Builder.toLazyByteStringWith (Builder.safeStrategy 1024 4096) mempty
+
+-- | Smart constructor for MerkleTree.
+fromFoldable :: (HasHash a, Foldable t) => t a -> MerkleTree a
+fromFoldable = fromList . F.toList
+
+-- | Smart constructor for MerkleTree.
+fromContainer :: (HasHash (Element t), Container t) => t -> MerkleTree (Element t)
+fromContainer = fromFoldable . toList
```

Comment:
Why do you use `fromFoldable`, but not `fromList` here? This way you basically do `toList` twice, it seems.